tests/task_queue_test.go

```diff
@@ -164,6 +167,129 @@ func (s *TaskQueueSuite) testTaskQueueRateLimitName(nPartitions, nWorkers int, u
+	s.OverrideDynamicConfig(dynamicconfig.MatchingNumTaskqueueWritePartitions, 1)
+	// Set a very low TTL for task queue info cache to ensure the rate limiter stats
+	// are refreshed frequently, avoiding stale data during the test.
+	s.OverrideDynamicConfig(dynamicconfig.TaskQueueInfoByBuildIdTTL, 1*time.Millisecond)
+
+	var (
+		mu       sync.Mutex
+		runTimes []time.Time
+	)
+
+	const (
+		workerRPS    = 50.0
+		drainTimeout = 30 * time.Second // Setting extremely high drain timeout to prevent flaking
+		activityName = "trackableActivity"
+	)
+
+	tv := testvars.New(s.T())
+	tv = tv.WithNamespaceName(namespace.Name(fmt.Sprintf("ns-rps-%.2f", apiRPS))) // To isolate state and avoid flaking
+
+	// Apply API rate limit on `activityTaskQueue`
+	_, err := s.FrontendClient().UpdateTaskQueueConfig(context.Background(), &workflowservice.UpdateTaskQueueConfigRequest{
+		Namespace:     s.Namespace().String(),
+		Identity:      tv.ClientIdentity(),
+		TaskQueue:     activityTaskQueue,
+		TaskQueueType: enumspb.TASK_QUEUE_TYPE_ACTIVITY,
+		UpdateQueueRateLimit: &workflowservice.UpdateTaskQueueConfigRequest_RateLimitUpdate{
+			RateLimit: &taskqueuepb.RateLimit{RequestsPerSecond: apiRPS},
+			Reason:    "Test API override",
+		},
+	})
+	s.NoError(err)
+
+	// Track activity run times
+	activityFunc := func(context.Context) error {
+		mu.Lock()
+		defer mu.Unlock()
+		runTimes = append(runTimes, time.Now())
+		return nil
+	}
+
+	workflowFn := func(ctx workflow.Context) error {
+		ao := workflow.ActivityOptions{
+			TaskQueue:           activityTaskQueue, // Route activities to a dedicated task queue named `activityTaskQueue`
+			StartToCloseTimeout: 5 * time.Second,
+		}
+		ctx = workflow.WithActivityOptions(ctx, ao)
+		return workflow.ExecuteActivity(ctx, activityName).Get(ctx, nil)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), drainTimeout)
+	defer cancel()
+
+	// Start the activity worker
+	activityWorker := worker.New(s.SdkClient(), activityTaskQueue, worker.Options{
+		TaskQueueActivitiesPerSecond: workerRPS,
+	})
+	activityWorker.RegisterActivityWithOptions(activityFunc, activity.RegisterOptions{Name: activityName})
+	s.NoError(activityWorker.Start())
+	defer activityWorker.Stop()
+
+	// Start the workflow worker
+	wfWorker := worker.New(s.SdkClient(), tv.TaskQueue().GetName(), worker.Options{})
+	wfWorker.RegisterWorkflow(workflowFn)
+	s.NoError(wfWorker.Start())
+	defer wfWorker.Stop()
+
+	// Launch workflows
+	for i := 0; i < taskCount; i++ {
+		_, err := s.SdkClient().ExecuteWorkflow(ctx, sdkclient.StartWorkflowOptions{
+			TaskQueue: tv.TaskQueue().GetName(),
+			ID:        fmt.Sprintf("wf-%d", i),
+		}, workflowFn)
+		s.NoError(err)
+	}
+
+	// Wait for all activities to complete
+	s.Eventually(func() bool {
+		mu.Lock()
+		defer mu.Unlock()
+		return len(runTimes) >= taskCount
+	}, drainTimeout, 100*time.Millisecond)
+
+	// Validate that API RPS was enforced
+	mu.Lock()
+	defer mu.Unlock()
```

Comment:
This should be unnecessary since you've already checked that all activities completed.